impeller/renderer/graphics_queue.h

```diff
@@ -0,0 +1,47 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_IMPELLER_RENDERER_GRAPHICS_QUEUE_H_
+#define FLUTTER_IMPELLER_RENDERER_GRAPHICS_QUEUE_H_
+
+#include <functional>
+
+#include "fml/status.h"
+#include "impeller/renderer/command_buffer.h"
+
+namespace impeller {
+
+/// @brief An interface for submitting command buffers to the GPU for
+///        encoding and execution.
+class GraphicsQueue {
+ public:
+  using CompletionCallback = std::function<void(CommandBuffer::Status)>;
+
+  GraphicsQueue() = default;
+
+  virtual ~GraphicsQueue() = default;
+
+  /// @brief Submit one or more command buffer objects to be encoded and
+  ///        executed on the GPU.
+  ///
+  ///        The order of the provided buffers determines the ordering in which
+  ///        they are submitted.
+  ///
+  ///        Optionally accepts a ccallback that will fire with an updated
+  ///        status based on encoding state. Only the Metal and Vulkan backend
```

Comment:
Perhaps a better way of putting this would be "The status only indicates if the command buffer was successfully submitted. Successful completion of the command buffer can only be checked in the completion callback.".

Reply:
So, if the status is not OK, can the caller expect a completion callback invocation? Otherwise they might have to perform cleanup themselves if the callback is dropped on the floor.

Reply:
Hmm, a better interface might be to guarantee that the callback always gets invoked but with CommandBuffer::Status::kError in case of failure to submit as well. This function can return void. This avoids burden on the caller to check for error and invoke the callback themselves in case the callback has cleanup they want to perform.

Reply:
That change would make it much more difficult to detect immediate submission errors